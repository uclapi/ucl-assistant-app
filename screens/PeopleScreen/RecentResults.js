// @flow
import PropTypes from "prop-types"
import React, { Component } from "react"
import { View } from "react-native"
import { connect } from "react-redux"
import { generate } from "shortid"

import { clearRecents as clearRecentsAction } from "../../actions/peopleActions"
import Button from "../../components/Button"
import SearchResult from "../../components/SearchResult"
import { CentredText, SubtitleText } from "../../components/Typography"

export class RecentResults extends Component {
  static mapDispatchToProps = (dispatch) => ({
    clearRecents: () => dispatch(clearRecentsAction()),
  })

  static mapStateToProps = (state) => ({
    recents: state.people.recents,
  })

  static propTypes = {
    clearRecents: PropTypes.func,
    navigation: PropTypes.shape().isRequired,
    recents: PropTypes.arrayOf(PropTypes.shape()),
  }

  static defaultProps = {
    clearRecents: () => { },
    recents: [],
  }

  viewPerson = (person) => () => {
    const { navigation } = this.props
    navigation.navigate(`PersonDetail`, person)
  }

  renderRecent = (recent = null) => {
    if (recent === null) {
      return null
    }
    return (
      <SearchResult
        key={generate()}
        topText={recent.name}
        bottomText={recent.department}
        type="person"
        buttonText="View"
        onPress={this.viewPerson(recent)}
      />
    )
  }

  render() {
    const { recents, clearRecents } = this.props
    if (recents.length === 0) {
      return null
    }
    return (
      <View>
        <SubtitleText>Recently Searched</SubtitleText>
        {recents.map(this.renderRecent)}
        {recents.length > 0 ? (
          <Button onPress={clearRecents}>Clear</Button>
        ) : (
            <CentredText>Recent results will appear here.</CentredText>
        )}
      </View>
    )
  }
}

export default connect(
  RecentResults.mapStateToProps,
  RecentResults.mapDispatchToProps,
)(RecentResults)
