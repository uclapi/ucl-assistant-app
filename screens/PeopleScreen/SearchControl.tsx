import PropTypes from "prop-types"
import React, { Component } from "react"
import { ActivityIndicator, StyleSheet, View } from "react-native"
import { connect } from "react-redux"
import { generate } from "shortid"

import {
  search as searchAction,
  searchClear as searchClearAction,
} from "../../actions/peopleActions"
import { SmallButton } from "../../components/Button"
import { Horizontal } from "../../components/Containers"
import { TextInput } from "../../components/Input"
import SearchResult from "../../components/SearchResult"
import { CentredText } from "../../components/Typography"

const styles = StyleSheet.create({
  textInput: {
    flex: 1,
    marginRight: 10,
  },
})

const MIN_QUERY_LENGTH = 4

export class SearchControl extends Component {
  static propTypes = {
    clear: PropTypes.func,
    error: PropTypes.string,
    isSearching: PropTypes.bool,
    navigation: PropTypes.shape().isRequired,
    search: PropTypes.func,
    searchResults: PropTypes.arrayOf(PropTypes.shape()),
    token: PropTypes.string,
  }

  static defaultProps = {
    clear: () => { },
    error: ``,
    isSearching: false,
    search: () => { },
    searchResults: [],
    token: ``,
  }

  static mapStateToProps = (state) => ({
    error: state.people.searchError,
    isSearching: state.people.isSearching,
    searchResults: state.people.searchResults,
    token: state.user.token,
  })

  static mapDispatchToProps = (dispatch) => ({
    clear: () => dispatch(searchClearAction()),
    search: (token, query) => dispatch(searchAction(token, query)),
  })

  static SEARCH_DELAY = 500;

  constructor(props) {
    super(props)
    this.searchTimer = null
    this.state = {
      query: ``,
    }
  }

  onQueryChange = (query: string) => {
    clearTimeout(this.searchTimer)
    this.searchTimer = setTimeout(
      () => this.search(query),
      SearchControl.SEARCH_DELAY,
    )
    this.setState({ query })
  }

  search = (query: string) => {
    const { search, token } = this.props
    search(token, query)
  }

  clear = (): void => {
    const { clearRecentResults } = this.props
    clearRecentResults()
    this.setState({ query: `` })
  }

  renderStatusText = () => {
    const { query } = this.state
    const { searchResults } = this.props
    if (query.length === 0) {
      return <CentredText>Start typing to get search results < /CentredText>
    }
    if (query.length < MIN_QUERY_LENGTH && searchResults.length === 0) {
      return <CentredText>Please enter a few more characters < /CentredText>
    }
    if (query.length > 0 && searchResults.length === 0) {
      return <CentredText>No results found.< /CentredText>
    }
    return null
  }

  viewPerson = (person) => () => {
    const { navigation } = this.props
    navigation.navigate(`PersonDetail`, person)
  }

  renderResult = (res = null) => {
    if (res === null) {
      return null
    }
    return (
      <SearchResult
        key= { generate() }
    topText = { res.name }
    bottomText = { res.department }
    type = "person"
    buttonText = "View"
    onPress = { this.viewPerson(res) }
      />
    )
  }

  render() {
    const { query } = this.state
    const {
      error,
      isSearching,
      searchResults,
    } = this.props
    return (
      <View>
      <Horizontal>
      <TextInput
            placeholder= "Search for a name or email..."
    onChangeText = { this.onQueryChange }
    value = { query }
    clearButtonMode = "always"
    style = { styles.textInput }
      />
    {
      query.length > 0 ? (
        <SmallButton onPress= { this.clear } > Clear < /SmallButton>
          ) : null}
      < /Horizontal>
    {
      error.length > 0 && query.length > 2 && (
        <CentredText>{`Error! ${error} `} </CentredText>
        )
  }

        {
  isSearching && <ActivityIndicator />}

  { this.renderStatusText() }

  { searchResults.map(this.renderResult) }
  </View>
    )
}
}

export default connect(
  SearchControl.mapStateToProps,
  SearchControl.mapDispatchToProps,
)(SearchControl)
