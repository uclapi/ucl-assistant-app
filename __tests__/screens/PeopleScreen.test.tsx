import "react-native"

import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import React from "react"
import renderer from "react-test-renderer"

import { RecentResults } from "../../screens/PeopleScreen/RecentResults"
import { SearchControl } from "../../screens/PeopleScreen/SearchControl"

Enzyme.configure({ adapter: new Adapter() })

describe(`RecentResults`, () => {
  it(`renders an empty RecentResults component`, async () => {
    const props = {
      clearRecents: jest.fn(),
      navigation: {
        navigate: jest.fn(),
      },
      recents: [],
    }
    const tree = renderer.create(<RecentResults {...props} />).toJSON()
    expect(tree).toBe(null)
  })

  it(`renders a filled RecentResults component`, async () => {
    const props = {
      clearRecents: jest.fn(),
      navigation: {
        navigate: jest.fn(),
      },
      recents: [
        {
          department: `Department of Agriculture`,
          name: `Mr Potato`,
        },
        {
          department: `Faculty of Medical Sciences`,
          name: `Chris P Bacon`,
        },
        {
          department: `Philosophy Department`,
          name: `Jeremy Bentham`,
        },
      ],
    }
    const tree = renderer.create(<RecentResults {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

it(`renders a SearchControl component`, async () => {
  const props = {
    clear: jest.fn(),
    error: ``,
    isSearching: false,
    navigation: {
      navigate: jest.fn(),
    },
    search: jest.fn(),
    searchResults: [],
    token: ``,
  }
  const component = shallow(<SearchControl {...props} />)
  expect(component).toMatchSnapshot()
})
