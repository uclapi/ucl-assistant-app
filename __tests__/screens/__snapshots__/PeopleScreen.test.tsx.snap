// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a SearchControl component 1`] = `
<View>
  <Horizontal
    bottomColour="#fefefe"
    contentContainerStyle={Object {}}
    keyboardAvoidingViewStyle={Object {}}
    mainTabPage={false}
    onRefresh={[Function]}
    refreshEnabled={false}
    refreshing={false}
    safeAreaViewStyle={Object {}}
    style={Object {}}
    topColour="#fefefe"
  >
    <TextInputComponent
      allowFontScaling={true}
      clearButtonMode="always"
      onChangeText={[Function]}
      placeholder="Search for a name or email..."
      rejectResponderTermination={true}
      style={
        Object {
          "flex": 1,
          "marginRight": 10,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </Horizontal>
  <CentredText
    style={Object {}}
  >
    Start typing to get search results 
  </CentredText>
</View>
`;
